extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,144 @@
+  )
+  {
+    this.inputRowSchema = inputRowSchema;
+    this.record = record;
+    this.headerParser = headerParser;
+    this.keyParser = keyParser;
+    this.payloadParser = payloadParser;
+    this.keyLabelPrefix = keyLabelPrefix;
+    this.recordTimestampLabelPrefix = recordTimestampLabelPrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    Map<String, Object> mergeList = new HashMap<>(this.headerParser.read());
+
+    if (this.keyParser != null) {
+      CloseableIterator<InputRow> keyIterator = this.keyParser.read();
+      // Key currently only takes the first row and ignores the rest.
+      if (keyIterator.hasNext()) {
+        MapBasedInputRow keyRow =  (MapBasedInputRow) keyIterator.next();
+        mergeList.put(
+            this.keyLabelPrefix + "key",
+            keyRow.getEvent().entrySet().stream().findFirst().get().getValue()
+        );
+      }
+      keyIterator.close();
+    }
+    // Add kafka record timestamp to the mergelist
+    mergeList.put(this.recordTimestampLabelPrefix + "timestamp", this.record.getRecord().timestamp());
+    //log.warn("LOKI Kafka record :" + new String(this.record.getRecord().value(), StandardCharsets.UTF_8));
+
+    CloseableIterator<InputRow> iterator = this.payloadParser.read();
+    List<InputRow> rows = new ArrayList<>();
+    try {
+      while (iterator.hasNext()) {
+      /* Currently we prefer payload attributes if there is a collision in names.
+          We can change this beahvior in later changes with a config knob. This default
+          behavior lets easy porting of existing inputFormats to the new one without any changes.
+       */
+        MapBasedInputRow row = (MapBasedInputRow) iterator.next();
+        Map<String, Object> event = new HashMap<>(row.getEvent());
+        mergeList.forEach((key, value) -> event.merge(key, value, (v1, v2) -> v1));
+
+        HashSet<String> newDimensions = new HashSet<String>(row.getDimensions());
+        newDimensions.addAll(mergeList.keySet());
+        // Free mergeList
+        mergeList = null;
+        // Remove the dummy timestamp added in KafkaInputFormat
+        newDimensions.remove("__kif_auto_timestamp");
+
+        final List<String> schemaDimensions = this.inputRowSchema.getDimensionsSpec().getDimensionNames();
+        final List<String> dimensions;
+        if (!schemaDimensions.isEmpty()) {
+          dimensions = schemaDimensions;
+        } else {
+          dimensions = Lists.newArrayList(
+              Sets.difference(newDimensions, this.inputRowSchema.getDimensionsSpec().getDimensionExclusions())
+          );
+        }
+        rows.add(new MapBasedInputRow(
+            this.inputRowSchema.getTimestampSpec().extractTimestamp(event),
+            dimensions,
+            event));
+      }
+    } catch (Exception e) {
+      log.error(e, "Encountered exception.");
+      if( e instanceof IOException ) {
+        throw (IOException)e;
+      }
```

Comment:
we should rethrow all exceptions, otherwise this will cause `RuntimeException`s to get swallowed and continue processing. I'm sure we have some examples in the code that calls input formats on how this is done